### ISSUE TITLE
inheritance for generics does not work

### ISSUE THREAD

**Issue by Araq** [nim-lang/Nim#88]
The following program should compile, but does not:

``` 'nimrod

type
  TGen[T] = object
    field: T

  TDerived[T] = object of TGen[T]
    nextField: T

proc doSomething[T](x: ref TGen[T]) =
  type
    Ty = ref TDerived[T]
  echo Ty(x).nextField

var
  x: ref TDerived[string]
new(x)
x.nextField = "test"

doSomething(x)
```

**Comment by ozra** [comment]
I stumbled upon this too. Pretty new to both Nim and, of course, the compilers source. 
@boydgreenfield @zah - If anyone is up for pair-programming over TeamView or similar - I'm interested. Two knuckle heads reasoning about the code. Have only tinkered in the lexer and the parser up til now though.

**Comment by BigBoneDaddy** [comment]
@ozra Yea it can be confusing to new comers

**Comment by ozra** [comment]
I could mention that I started debugging this one, but I simply can't make the time to finish it. Some one more clever or with more time is needed. If someone does, and any of what I've looked at can be of help (unsure), just ping me.

**Comment by BigBoneDaddy** [comment]
@ozra Can you share your infos and findings with us here?

**Comment by dom96** [comment]
@ozra pair hacking on Nim sounds like a fun idea

**Comment by BigBoneDaddy** [comment]
@ozra @dom96 Just a couple of burly men together hacking away on Nim making it the best it can be. What is better than that?!

**Comment by dom96** [comment]
@BigBoneDaddy "burly"?

**Comment by ozra** [comment]
@BigBoneDaddy Well, to be honest, not much :-/. I figured if someone stumbled on something, I perhaps, maybe, possibly, could have an idea of some sort from my rumbling arounds.

I got it to match the type, but then the type vars didn't match, or I got it to work a bit further with another solution, but then there'd be an error in GC optimized collectors generation. So, I didn't really fully get a grip on the type graph, what should be and what shouldn't. What role the 'type-id's play and what their rules for uniqueness are. What's a miss and what's intended. It's not exactly documented at bible verse length... So, I finally had to give up (put food on the table kind of work and the garden takes their hours.)

**Comment by BigBoneDaddy** [comment]
@dom96 Yes burly means big and strong and robust

**Comment by ozra** [comment]
Haha, @BigBoneDaddy, I don't fit that description - except strong and robust perhaps, haha. Anyway, if you take on this bug it would be super, and I could try to make as much time as possible to help when possible.

**Comment by def-** [comment]
Are you sure this is fixed? #4398 says "This doesn't finally fix #88"

**Comment by Araq** [comment]
Newer version of this bug: 

```nim


type
  TGen[T] = object of RootObj
    field: T

  TDerived[T] = object of TGen[T]
    nextField: T

proc doSomething[T](x: ref TGen[T]) =
  type
    Ty = ref TDerived[T]
  echo Ty(x).nextField

var
  x: ref TDerived[string]
new(x)
x.nextField = "test"

doSomething(x)
```
